### DISCUSSION TITLE
Avoiding static arguments when using JAX for MARL

### DISCUSSION THREAD

**Discussion by cool-RR** [jax-ml/jax#23200] [Q&A]
Hi folks!

I'm using JAX for MARL experiments using an opponent shaping algorithm called [POLA](https://r.rachum.com/pola-code). This algorithm is complicated and takes a long time to compile, between minutes and 12 hours, depending on options. I believe that one of the causes for how long the compilations are taking is that I have many jitted functions that get an agent index as an argument, and that argument is marked as static.

Here is a grossly simplified example of code that I hope helps explain the issue: https://gist.github.com/cool-RR/dd0b548c1990ec111e03c5b8e0863808

The `Population` class has multiple agents, i.e. multiple `TrainState` objects in a list. The `update_agent` method trains one agent while depending on the behavior of all the other agents. As you can see on line 38, the `agent` argument is marked as static. If I don't mark it as static, JAX would raise an exception, since I'm using `agent` as an array index later.

This is a simplified example; in my real code I have a dozen different methods that call each other and they all have an `agent` argument that's marked as static.

**Am I right in my intution that this static argument is making the compilation of my program slower? Is there any way for me to make it so the `agent` argument would not be static?**

Thanks for your help,
Ram Rachum.

**Comment by jakevdp** [comment]
As written, you will have to re-compile your `update_agent` for each unique value of the integer `agent` (JIT compilation caches are keyed based on the static arguments). This will likely cause a significant slowdown if you're working with a large number of agents.

A better pattern would be to avoid having a list of train states, but instead have a single batched train state which you can index into using a dynamic agent index. In other words, use a *struct-of-arrays* pattern for your train states rather than the current *list-of-structs* pattern. A further benefit of this would be that you could replace the `for` loop in `loss_fun` with a `vmap` followed by some post-processing of the outputs, which would speed up the compilation and runtime even further.

Does that make sense?

**Reply by cool-RR** [reply]
Thank you for the quick reply.

I'm interested in the pattern you propose but I don't understand what you mean exactly. Here are a couple of confusions I have: 

1. You say I should use _struct-of-arrays_ rather than _list-of-structs_. But a `TrainState` is already a struct with a pytree of `params` in it , so how can I do that? Are you suggesting that I have one big `TrainState` which has an extra first axis on each array of its `params`, so it implicitly contains multiple train states?
2. I thought that when you use `vmap` to vectorize a function, the different items in the batch are computed independently of each other and can't really depend on each other. I do need my agents to affect each other, like in the example.


Thanks for your help,
Ram.

**Reply by jakevdp** [reply]
> Are you suggesting that I have one big TrainState which has an extra first axis on each array of its params, so it implicitly contains multiple train states?

Exactly.

> I thought that when you use vmap to vectorize a function, the different items in the batch are computed independently of each other and can't really depend on each other.

This is true

> I do need my agents to affect each other, like in the example.

The example you gave can be expressed in terms of a nested `vmap` I think – it has an outer-apply structure, where you'd compute an NxN grid of results for N train states, zero-out the diagonal, and then aggregate them at the end.

**Reply by cool-RR** [reply]
> > I do need my agents to affect each other, like in the example.
> 
> The example you gave can be expressed in terms of a nested `vmap` I think – it has an outer-apply structure, where you'd compute an NxN grid of results for N train states, zero-out the diagonal, and then aggregate them at the end.

Hmm, I don't understand what you mean, but the question is whether this trick could be applied to my real code. (The example I gave is simplified.)

In opponent shaping algorithms, each agent is trained to maximize its reward not against the current behavior of the other agents but against how these agents would be trained in response to the agent's new training. (It can be a difficult concept to wrap one's mind around.) Another way to say it is that agents are differentiating through the other agents' parameters.

Do you think that this trick could be applied to the system I described? Or could there be another way to avoid a separate compilation for each agent?

Or also... if there's a separate compilation per agent, is it at least possible to do these compilations in parallel? Because right now the compilation (which includes gradients) is always taking 100% of one CPU core while it's running for a long time.

**Reply by jakevdp** [reply]
I think it's definitely possible to express that algorithm without static indices.

Write the computation you need for a single agent, then `vmap` it across all agents, and remove/zero-out the value for the agent with the same index.

**Reply by cool-RR** [reply]
Thank you. I appreciate the time you're putting towards helping me and I'm trying to make the most out of it.

I've been mulling over how to implement what you suggest. For starters, I had trouble vmapping across different train states ([example code](https://gist.github.com/cool-RR/1a017ee38b6521d985832e527fc70181)). I get this exception:

```
builtins.ValueError: vmap was requested to map its argument along axis 0, which implies that its rank should be at least 1, but is only 0 (its shape is ())
```

So I'm guessing that instead of a list of `TrainState` objects I should build the `TrainState` megazord that we previously discussed and then feed it to `vmap`. I've used this function to stack multiple train states:

```
def stack_pytrees(pytrees):
    return jax.tree.map(lambda *arrays: jnp.stack(arrays), *pytrees)
```

However, when I run [the code](https://gist.github.com/cool-RR/013348bfdfb3a65b49dd0fe0a6586476) I get this exception: 

```
(truncated)
File "C:\Users\Administrator\.venvs\polina_env\Lib\site-packages\jax\_src\tree_util.py", line 342, in tree_map
  all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]

builtins.ValueError: Mismatch custom dataclass node data: (<bound method Module.apply of SimpleNN()>, GradientTransformationExtraArgs(init=<function chain.<locals>.init_fn at 0x0000022ABBBED800>, update=<function chain.<locals>.update_fn at 0x0000022ABBBED6C0>)) != (<bound method Module.apply of SimpleNN()>, GradientTransformationExtraArgs(init=<function chain.<locals>.init_fn at 0x0000022ABBC089A0>, update=<function chain.<locals>.update_fn at 0x0000022ABBC08900>)); value: TrainState(step=0, apply_fn=<bound method Module.apply of SimpleNN()>, params={'Dense_0': {'kernel': Array([[-0.27713674, 
 ... (truncated)
```

If I understand correctly, the fields that are mentioned in the exception are those that are marked as `pytree_node=False`, which I guess JAX expects to be identical.

**How do I deal with this problem? Am I doing something wrong?**

**Reply by cool-RR** [reply]
I spent a couple more hours trying different things. Did you mean that I'd use `jax.flatten_util.ravel_pytree` to turn a `TrainState` into one big array?

**Reply by jakevdp** [reply]
You could define your batched training states using something like this:
```python
train_states = jax.vmap(create_train_state, axis_size=n_agents)()
```
and then your loss function with dynamic `agent` index could look like this:
```python
def loss_fn(params, agent=agent, x=x):
    # Example: compute loss based on all other agents
    outputs = jax.vmap(train_states.apply_fn, in_axes=(0, None))(train_states.params, x)
    agent_output = outputs[agent]
    other_agent_mask = jnp.arange(n_agents)[:, None] != agent
    return jnp.mean((agent_output - jnp.mean(outputs, where=other_agent_mask))**2)
```
that's more-or-less the idea. What do you think?

**Reply by cool-RR** [reply]
Oh, I didn't get a notification about your reply because of a temporary email problem and I ended up spending the last couple of days completely revamping my code to store the train states as a raveled stack. I didn't use `vmap`; because I moved from maintaining a list of train states to maintaining an array which is a stack of raveled train states, I could just index freely on this array without JAX complaining about using the `__index__` method.

After fixing dozens of bugs, I ran it to completion. It took 3.5 times longer to compile than the original version 😭

I wonder whether the compilation is so slow because of the raveling and unraveling. I don't know how to check that. I did try to compile a sample function that just ravels and unravels a train state a million times and it compiles almost instantly. I don't know whether I can draw any conclusions from that. I hope I won't have to scrap the work I did these last two days. Any help would be appreciated.

Regarding your `vmap` suggestion: It seems to fit well for the code sample I gave, but I don't know whether it could work for my actual code. In my codebase I have several places where I need to pick and choose train states. For example, I sometimes need to build a population that has all the train states of an existing population, except that train state 7 is replaced with a new train state. `vmap` can't do that for me, right? How can I access the 7th item of the train state that contains multiple train states? I suppose I could use `jax.tree.map` to do that, but I wonder if that'll result in the same slowness that I've gotten when using `pytree_ravel`.

**Reply by jakevdp** [reply]
I wouldn't be surprised if `pytree_ravel` is causing slowness, because any time you're working with lists of structs rather than structs of arrays, you're going to end up creating a lot of separate array objects.

Regarding selecting just the 7th train state from a batched train state: You should be able to do this by tree-mapping the indexing operation. That should be possible to do within a `vmap` without any problem.

Without seeing your code I can only speculate, but if you're using `pytree_ravel` to switch back and forth between lists-of-structs and structs-of-arrays, I suspect you'll end up with a slow solution. You should try to use the structs-of-arrays pattern from the beginning if possible.

**Reply by cool-RR** [reply]
I've now shared my code with you. The `master` branch has the old version of the code before this change, and the `no-static` branch has the new version which is a single commit. It has lots of infrastructure changes though, to support this change, so it probably won't be practical to read through it. 

> I wouldn't be surprised if `pytree_ravel` is causing slowness, because any time you're working with lists of structs rather than structs of arrays, you're going to end up creating a lot of separate array objects.

Why would `pytree_ravel` imply lists of structs? I put a train state in it, and then it converts it to a big flat array. Where are the lists here?

Some of the changes I made in this commit are changes from lists-of-structs to structs of arrays, like in the class `StepStack` which used to have a list of structs of arrays, and now has just arrays with no lists. And still the compilation is slower.

**Reply by cool-RR** [reply]
And also, you suggested tree-mapping indexing. Wouldn't that be as slow as the raveling-and-unraveling method?

**Reply by jakevdp** [reply]
Sorry, I must have misunderstood what you meant when you mentioned `pytree_ravel` (I assume what you're referring to is `jax.flatten_util.ravel_pytree`)? I'll take a look at your code, but I'm not sure I'll have the bandwidth to dig into diffs between your branches

Tree map indexing is not as slow as ravelling and unravelling: it's just applying indexing to the batched arrays within the training data structure.

**Reply by jakevdp** [reply]
Sorry, I opened your repository, and I have no idea where I'm supposed to be looking. There are dozens of files there.

**Reply by jakevdp** [reply]
General compilation time debugging tips: any time you're using Python control flow (e.g. a `for` loop, a `while` loop, or recursive function calls) you're likely generating many XLA operations in sequence, because JAX flattens all Python control flow. The more XLA operations you have, the longer your program will be, and the longer your program is, the longer it will take to compile (compile time is generally super-linear in the number of operations)

So to fix slow compilation, first look for places where you have `for` loops within a `jit` compiled function, and either replace them with `vmap`-based batching if the operations have no dependence between batching, or with `fori_loop`/`scan` in cases where operations have sequential dependence.

This is why my first recommendation to you was to avoid the list-of-structs (which require looping via Python control flow) in favor of struct-of-arrays (which allow batching via `vmap`).

**Reply by jakevdp** [reply]
> Yes, it's pretty big. [pola_algorithm.py](https://github.com/cool-RR/polina/blob/no-static/polina/pola_algorithm.py) is probably the most important file. The `Population.train` method there is kind of the main function. The `Population` class used to contain a field `brains` which was a tuple of `Brain` objects. (A `Brain` object has a pair of train states, one for actor and one for critic, and a bunch of metadata.) After the change, the `Population` class has an attribute `raveled_brain_by_agent` which is an array, and then `Population.__getitem__` returns `self.brain_unraveler(self.raveled_brain_by_agent[i])` so you could get brains by doing `population[agent]`.

Yeah - it looks like you have thousands of lines of existing code, and it seems like that may be tricky to refactor the fundamental data structures used in that code. But the general advice above will apply: avoid Python control flow if you find that compilation times are getting out of hand.

**Reply by cool-RR** [reply]
I understand, thank you. I'll think on how to proceed.

**Comment by cool-RR** [comment]
Yes, it's pretty big. [pola_algorithm.py](https://github.com/cool-RR/polina/blob/no-static/polina/pola_algorithm.py) is probably the most important file. The `Population.train` method there is kind of the main function. The `Population` class used to contain a field `brains` which was a tuple of `Brain` objects. (A `Brain` object has a pair of train states, one for actor and one for critic, and a bunch of metadata.) After the change, the `Population` class has an attribute `raveled_brain_by_agent` which is an array, and then `Population.__getitem__` returns `self.brain_unraveler(self.raveled_brain_by_agent[i])` so you could get brains by doing `population[agent]`.
